ml_tools/utils/plotting.py

```diff
@@ -313,40 +320,154 @@ def plot_ice_pdp(models:          Dict[str, PredictionStrategy],
                      estimator=None, alpha=0.3, legend=False, linewidth=1)
 
-        # PDP Curve
         pdp_df = ice_df.groupby('value')['prediction'].mean().reset_index()
         sns.lineplot(data=pdp_df, x='value', y='prediction', color='red', label='PDP', linewidth=2)
 
-        # Rug plot for original (unperturbed) values
         sns.rugplot(x=values, height=0.03, color='black', alpha=0.3)
 
-        # Axis labels
         predicted_feature_label = predicted_feature if predicted_feature_label is None else predicted_feature_label
         plt.xlabel(input_feature)
         plt.ylabel(predicted_feature_label)
         plt.grid(True)
         plt.tight_layout()
         plt.savefig(f"{fig_name_prefix}_{label}_{input_feature}.png", dpi=600, bbox_inches='tight')
         plt.close()
+
+
+
+def plot_shap(models:          Dict[str, PredictionStrategy],
+              state_series:    List[StateSeries],
+              feature_plots:   Dict[str, List[str]],
+              algorithm:       str = 'auto',
+              fig_name_prefix: str = 'shap',
+              state_index:     int = -1,
+              array_index:     int = 0,
+              num_samples:     int = 50,
+              num_procs:       int = 1) -> None:
+    """ Function to plot SHAP feature importance summary for a given set of models.
+
+    Parameters
+    ----------
+    models : Dict[str, PredictionStrategy]
+        The collection of models (i.e. prediction strategies) whose sensitivities are to be plotted.
+        The dictionary key will be the suffix of the figure file name
+    state_series : List[StateSeries]
+        The state series to use for plotting
+    feature_plots : Dict[str, List[str]]
+        Specification for what sets of input features to make SHAP plots for. The key corresponds to the
+        set label for file naming, and the value the list of input features the corresponding plot will include
+    algorithm : str
+        The SHAP algorithm to use, either 'auto', 'permutation', or 'partition' (Default: 'auto')
+    fig_name_prefix : str, optional
+        The prefix for the figure files that will be created (Default: 'shap')
+    state_index : int, optional
+        The index of the state in the series to be analyzed (Default: -1, last state).
+    array_index : int
+        The index of the predicted value array to be plotted (Default: 0)
+    num_samples : int, optional
+        The number of sampled points for assessing the SHAP values. (Default: 50)
+    num_procs : int
+        The number of parallel processors to use when performing the SHAP evaluation (Default: 1)
+    """
+
+    state_series       = random.sample(state_series, min(num_samples, len(state_series)))
+    df                 = series_to_pandas(state_series)
+    all_input_features = list(df.columns)
+    X                  = df.to_numpy(dtype=float)
+
+    def compute_shap(args) -> np.ndarray:
+        """ Function for computing shap values in parallel.
+        """
```

Comment:
Do you want to document `args` and the return value here?

Reply:
nah, these are super lightweight helper functions strictly for the `ThreadPoolExecutor`.  It should be pretty obvious what's happening.

Reply:
lol, famous last words...